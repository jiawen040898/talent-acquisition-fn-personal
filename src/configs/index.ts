export * from './auth.config';
export * from './aws.config';
export * from './skill-extract-api.config';
export * from './unleash.config';
