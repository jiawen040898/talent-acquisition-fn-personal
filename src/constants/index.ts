export * from './action-history.enum';
export * from './auth-error.enum';
export * from './cache-store.enum';
export * from './candidate.event-type.enum';
export * from './channel.enum';
export * from './contact-type.enum';
export * from './ds-lookup-api-errors.enum';
export * from './email-event.enum';
export * from './email-type.enum';
export * from './error-msg.enum';
export * from './errors';
export * from './feature-toggle-name.enum';
export * from './fit-score-recipe.enum';
export * from './identity-event.enum';
export * from './job.enum';
export * from './job-application-score-type.enum';
export * from './job-application-skill.enum';
export * from './job-application-status.enum';
export * from './person-score.enum';
export * from './person-score-event.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework.enum';
export * from './s3-errors.enum';
export * from './screening-answer-criteria.enum';
export * from './skill-extract-api-errors.enum';
export * from './talent-acquisition-event.enum';
export * from './talent-management-event.enum';
export * from './timeline.enum';
export * from './unleash-errors.enum';
export * from './user.enum';
