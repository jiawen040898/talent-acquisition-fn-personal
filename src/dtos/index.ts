export * from './audit-data.dto';
export * from './candidate-event.dto';
export * from './create-job.dto';
export * from './fit-score-recipe.dto';
export * from './geolocation.dto';
export * from './identity-update-user-event.dto';
export * from './job-application.dto';
export * from './job-application-action-history.dto';
export * from './job-application-attachment.dto';
export * from './job-application-career.dto';
export * from './job-application-contact.dto';
export * from './job-application-education.dto';
export * from './job-application-questionnaire.dto';
export * from './job-application-resume.dto';
export * from './job-application-score.dto';
export * from './job-application-screening-answer.dto';
export * from './person-score-event.dto';
export * from './place.dto';
export * from './talent-acquisition-event.dto';
export * from './talent-management-event.dto';
