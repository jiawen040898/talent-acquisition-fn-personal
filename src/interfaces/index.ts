export * from './api.interface';
export * from './api-response.interface';
export * from './aws.interface';
export * from './candidate.interface';
export * from './cognitive-score.interface';
export * from './email.interface';
export * from './event-model.interface';
export * from './job-application-score.interface';
export * from './pairwise-score.interface';
export * from './personality.interface';
export * from './personality-score.interface';
export * from './skill-extract-api.interface';
export * from './user.interface';
