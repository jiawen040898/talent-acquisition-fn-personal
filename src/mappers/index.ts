export * from './job-application-mapper';
export * from './job-application-score-mapper';
