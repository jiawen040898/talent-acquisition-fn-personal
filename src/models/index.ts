export * from './audit-data.entity';
export * from './entities';
export * from './job.entity';
export * from './job-application.entity';
export * from './job-application-action-history.entity';
export * from './job-application-attachment.entity';
export * from './job-application-career.entity';
export * from './job-application-contact.entity';
export * from './job-application-education.entity';
export * from './job-application-label.entity';
export * from './job-application-questionnaire.entity';
export * from './job-application-resume.entity';
export * from './job-application-score.entity';
export * from './job-application-screening-answer.entity';
export * from './job-distribution-score.entity';
export * from './job-questionnaire.entity';
export * from './job-screening-question.entity';
export * from './job-user-access.entity';
