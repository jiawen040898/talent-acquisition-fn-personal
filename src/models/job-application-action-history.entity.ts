import {
    Column,
    Entity,
    JoinColumn,
    ManyToOne,
    PrimaryGeneratedColumn,
} from 'typeorm';

import { AuditDataEntity } from './audit-data.entity';
import { JobApplication } from './job-application.entity';

@Entity({ name: 'job_application_action_history' })
export class JobApplicationActionHistory extends AuditDataEntity {
    @PrimaryGeneratedColumn()
    id?: number;

    @Column({
        type: 'uuid',
    })
    job_application_id!: string;

    @Column({
        type: 'uuid',
    })
    action_type!: string;

    @Column('simple-json', {
        nullable: true,
    })
    value?: JSON;

    @Column({
        length: 255,
    })
    created_username!: string;

    @ManyToOne(
        () => JobApplication,
        (jobApplication) => jobApplication.action_histories,
    )
    @JoinColumn({ name: 'job_application_id' })
    job_application?: JobApplication;
}
