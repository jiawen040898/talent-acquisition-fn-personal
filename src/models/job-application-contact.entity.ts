import { ContactType } from '@pulsifi/constants';
import {
    Column,
    Entity,
    JoinColumn,
    ManyToOne,
    PrimaryGeneratedColumn,
} from 'typeorm';

import { AuditDataEntity } from './audit-data.entity';
import { JobApplication } from './job-application.entity';

@Entity({ name: 'job_application_contact' })
export class JobApplicationContact extends AuditDataEntity {
    @PrimaryGeneratedColumn()
    id?: number;

    @Column({
        type: 'uuid',
    })
    job_application_id!: string;

    @Column({
        enum: ContactType,
        length: 20,
    })
    contact_type!: string;

    @Column({
        length: 5000,
    })
    value!: string;

    @Column()
    is_primary!: boolean;

    @ManyToOne(
        () => JobApplication,
        (jobApplication) => jobApplication.contacts,
    )
    @JoinColumn({ name: 'job_application_id' })
    job_application?: JobApplication;
}
