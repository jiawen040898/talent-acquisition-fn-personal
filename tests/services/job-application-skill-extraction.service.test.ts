import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { sdkStreamMixin } from '@aws-sdk/util-stream-node';
import { snsService } from '@pulsifi/fn';
import { JobApplication } from '@pulsifi/models';
import {
    FeatureToggleService,
    JobApplicationSkillExtractionService,
    SkillExtractAPIService,
} from '@pulsifi/services';
import { mockClient } from 'aws-sdk-client-mock';
import { Readable } from 'stream';
import { DataSource, Repository } from 'typeorm';

import * as database from '../../src/database';
import sampleFailedResume from '../samples/sample-failed-resume.json';
import sampleResume from '../samples/sample-resume.json';
import { getTestDataSourceAndAddData } from '../setup';
import { testData } from './fixtures/test-data';

jest.mock('@pulsifi/fn/services/sns.service');
const mockFeatureToggleService = jest.fn();
const mockSkillExtractAPIResponse = jest.fn();

describe('JobApplicationSkillExtractionService', () => {
    const jobApplicationSkillExtractionService =
        JobApplicationSkillExtractionService;
    let dataSource: DataSource;
    let jobApplicationRepo: Repository<JobApplication>;
    const featureToggleService = FeatureToggleService;
    const skillExtractAPIService = SkillExtractAPIService;

    beforeAll(async () => {
        dataSource = await getTestDataSourceAndAddData(
            testData.entitiesToBeAdded,
        );

        jest.spyOn(database, 'getDataSource').mockImplementation(() => {
            return Promise.resolve(dataSource);
        });

        jobApplicationRepo = dataSource.getRepository(JobApplication);
        featureToggleService.isUnleashFlagEnabled = mockFeatureToggleService;
        skillExtractAPIService.getOpenAISkills = mockSkillExtractAPIResponse;
    });

    describe('extractJobApplicationSkills', () => {
        const s3ClientMock = mockClient(S3Client);

        afterEach(async () => {
            jest.mocked(snsService.send).mockReset();
            s3ClientMock.reset();
        });

        it('should return void if no resume and role fit', async () => {
            // Act
            const result =
                await jobApplicationSkillExtractionService.extractJobApplicationSkills(
                    1,
                    testData.jobApplicationPayload3,
                );

            // Assert
            expect(result).toBeUndefined();
        });

        describe('if have candidate skills input', () => {
            it('should successfully extract and save by prioritize candidate skills if they match the candidate skills.', async () => {
                // Arrange
                mockFeatureToggleService.mockResolvedValueOnce(true);
                mockSkillExtractAPIResponse.mockResolvedValueOnce(
                    testData.mockSkillExtractAPIResponseWithoutResume,
                );

                // Act
                await jobApplicationSkillExtractionService.extractJobApplicationSkills(
                    1,
                    testData.jobApplicationPayload,
                );

                // Assert
                const jobApplication = await jobApplicationRepo.findOne({
                    where: { id: testData.jobApplication.id },
                });

                expect(jobApplication!.skills).toMatchSnapshot();
            });

            it('should successfully extract and save candidate skills, including those not matching the candidate skills.', async () => {
                // Arrange
                mockFeatureToggleService.mockResolvedValueOnce(true);
                mockSkillExtractAPIResponse.mockResolvedValueOnce({
                    ...testData.mockSkillExtractAPIResponseWithoutResume,
                    competencies: [{ skill: 'NodeJS', category: 'Softwares' }],
                });

                // Act
                await jobApplicationSkillExtractionService.extractJobApplicationSkills(
                    1,
                    testData.jobApplicationPayload,
                );

                // Assert
                const jobApplication = await jobApplicationRepo.findOne({
                    where: { id: testData.jobApplication.id },
                });

                expect(jobApplication!.skills).toMatchSnapshot();
            });
        });

        describe('if no candidate skills input', () => {
            it('should successfully extract and save all unique daxtra & openAI skills', async () => {
                // Arrange
                const s3ClientMock = mockClient(S3Client);
                const stream = new Readable();
                stream.push(JSON.stringify(sampleResume));
                stream.push(null);
                const sdkStream = sdkStreamMixin(stream);
                s3ClientMock.on(GetObjectCommand).resolves({ Body: sdkStream });

                mockFeatureToggleService.mockResolvedValueOnce(true);
                mockSkillExtractAPIResponse.mockResolvedValueOnce(
                    testData.mockSkillExtractAPIResponse,
                );

                // Act
                await jobApplicationSkillExtractionService.extractJobApplicationSkills(
                    1,
                    testData.jobApplicationPayload2,
                );

                // Assert
                const jobApplication = await jobApplicationRepo.findOne({
                    where: { id: testData.jobApplication2.id },
                });

                expect(jobApplication!.skills).toMatchSnapshot();
            });

            it('should return empty extraction result , if daxtra text resume has no content', async () => {
                // Arrange
                const stream = new Readable();
                stream.push(JSON.stringify(sampleFailedResume));
                stream.push(null);
                const sdkStream = sdkStreamMixin(stream);
                s3ClientMock.on(GetObjectCommand).resolves({ Body: sdkStream });

                mockFeatureToggleService.mockResolvedValueOnce(true);
                mockSkillExtractAPIResponse.mockResolvedValueOnce(
                    testData.mockSkillExtractAPIEmptyResponse,
                );

                // Act
                await jobApplicationSkillExtractionService.extractJobApplicationSkills(
                    1,
                    testData.jobApplicationPayload4,
                );

                // Assert
                const jobApplication = await jobApplicationRepo.findOne({
                    where: { id: testData.jobApplication4.id },
                });

                expect(jobApplication!.skills).toMatchSnapshot();
            });
        });
    });
});
