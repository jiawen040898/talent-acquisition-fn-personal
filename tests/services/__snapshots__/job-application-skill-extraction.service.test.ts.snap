// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobApplicationSkillExtractionService extractJobApplicationSkills if have candidate skills input should successfully extract and save by prioritize candidate skills if they match the candidate skills. 1`] = `
[
  {
    "name": "Python",
    "proficiency": "expert",
    "source": "candidate",
  },
]
`;

exports[`JobApplicationSkillExtractionService extractJobApplicationSkills if have candidate skills input should successfully extract and save candidate skills, including those not matching the candidate skills. 1`] = `
[
  {
    "name": "Python",
    "proficiency": "expert",
    "source": "candidate",
  },
  {
    "name": "NodeJS",
    "proficiency": null,
    "source": "openai",
  },
]
`;

exports[`JobApplicationSkillExtractionService extractJobApplicationSkills if no candidate skills input should return empty extraction result , if daxtra text resume has no content 1`] = `[]`;

exports[`JobApplicationSkillExtractionService extractJobApplicationSkills if no candidate skills input should successfully extract and save all unique daxtra & openAI skills 1`] = `
[
  {
    "name": "Organisational Skills",
    "proficiency": null,
    "source": "openai",
  },
  {
    "name": "Product Launch",
    "proficiency": "expert",
    "source": "daxtra",
  },
  {
    "name": "Employee Engagement",
    "proficiency": "expert",
    "source": "daxtra",
  },
  {
    "name": "Competitor Analysis",
    "proficiency": null,
    "source": "daxtra",
  },
  {
    "name": "Cost Estimating",
    "proficiency": null,
    "source": "daxtra",
  },
  {
    "name": "Delegation Skills",
    "proficiency": null,
    "source": "daxtra",
  },
  {
    "name": "Managerial Skills > People Management",
    "proficiency": null,
    "source": "daxtra",
  },
  {
    "name": "Operations Management",
    "proficiency": null,
    "source": "daxtra",
  },
  {
    "name": "Teamwork",
    "proficiency": null,
    "source": "daxtra",
  },
]
`;
